File: cpp/_127.cpp
// O(N) / O(N)
class Solution {
public:
    int ladderLength(string begin_word, string end_word, vector<string>& word_list) {
        
        // construct word map
        unordered_map<string, vector<string>> word_map;
        for (const string &word : word_list) {
            for (size_t i = 0; i < word.size(); ++i) {
                string tmp_word = word;
                tmp_word[i] = '*';
                word_map[tmp_word].push_back(move(word));
            }
        }
        
        // prep bfs
        unordered_set<string> seen;
        queue<string> q;
        q.push(begin_word);
        int level = 0;
        
        // bfs
        while (!q.empty()) {
            int level_size = q.size();
            level++;
            for (int i = 0; i < level_size; ++i) {
                const string curr_word = q.front();
                q.pop();
                for (int j = 0; j < curr_word.size(); ++j) {
                    string tmp_word = curr_word;
                    tmp_word[j] = '*';
                    auto it = word_map.find(tmp_word);
                    if (it == word_map.cend())
                        continue;
                    for (const string &word : it->second) {
                        if (word == end_word)
                            return level + 1;
                        if (seen.count(word))
                            continue;
                        q.push(word);
                        seen.insert(word);
                    }
                }
            }
        }
        
        return 0;
    }
};

File: python/_127.py
# Solution 1. BFS
# O(N) / O(N)
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        
        # construct word dict
        word_len = len(beginWord)
        word_dict = collections.defaultdict(list)
        for word in wordList:
            for i in range(word_len):
                word_dict[word[:i] + '*' + word[i+1:]].append(word)
                
        # bfs prep
        queue = collections.deque()
        queue.append(beginWord)
        seen = set()
        level = 0
        
        # bfs
        while queue:
            level_size = len(queue)
            level += 1
            for _ in range(level_size):
                curr_word = queue.popleft()
                for i in range(word_len):
                    temp = curr_word[:i] + '*' + curr_word[i+1:]
                    for word in word_dict[temp]:
                        if endWord == word:
                            return level + 1
                        if word not in seen:
                            seen.add(word)
                            queue.append(word)
        
        return 0


# Solution 2. two-way BFS
# O(N) / O(N)
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        
        # construct word dict
        word_len = len(beginWord)
        word_dict = collections.defaultdict(list)
        for word in wordList:
            for i in range(word_len):
                word_dict[word[:i] + '*' + word[i+1:]].append(word)
                
        # bfs prep
        queue = collections.deque()
        queue.append(beginWord)
        queue.append(endWord)
        seen = {beginWord: 1, endWord: -1}
        
        # bfs
        while queue:
            cur_word = queue.popleft()
            level = seen[cur_word]
            for i in range(word_len):
                temp = cur_word[:i] + '*' + cur_word[i+1:]
                for word in word_dict[temp]:
                    if word not in seen:
                        seen[word] = level + 1 if level > 0 else level - 1
                        queue.append(word)
                    if (seen[word] > 0) ^ (level > 0):
                        return abs(seen[word] - level)
                    
        return 0